### ISSUE TITLE
FullChainFile chain may be an empty list

### ISSUE THREAD

**Issue by zenhack** [zenhack/simp_le#23]
<a href="https://github.com/dracos"><img src="https://avatars.githubusercontent.com/u/154364?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [dracos](https://github.com/dracos)**
_Wednesday Apr 06, 2016 at 17:38 GMT_
_Originally opened as https://github.com/kuba/simp_le/issues/95_

----

https://github.com/kuba/simp_le/blob/8f258bc098a84b7a20c2732536d0740244d814f7/simp_le.py#L709 tests if `data.chain` is None, but its parent's `load_from_content` means that if e.g. the file exists but is empty or corrupt, it will be an empty list. I think the check should also check for the empty list as well as None.

```
$ touch foo.pem
$ python
>>> import simp_le
>>> simp_le.FullChainFile(path="foo.pem").load()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "simp_le.py", line 713, in load
    cert, chain = data.chain[0], data.chain[1:]
IndexError: list index out of range
```

**Comment by buchdag** [comment]
Tested this out of curiosity, turns out there is almost nothing done to handle empty or corrupt files. 

I'm working on this, seems the original idea (linked above)  was to consider unreadable data as no data at all and let `simp_le` overwrite the existing empty/corrupt files with new, clean data. Seems simple and sound to me, maybe with an additional warning ?

**Comment by zenhack** [comment]
I don't like the idea of just silently ignoring/overwriting a corrupt file. That probably shouldn't happen with any regularity and I think it's best to just report it to the user as a hard error.

Probably the same thing for empty files.

**Comment by buchdag** [comment]
So catch and specific error ?

**Comment by zenhack** [comment]
Yeah, that was my thinking.

**Comment by zenhack** [comment]
Quoting Nicolas Duchon (2017-12-11 18:45:59)

>    So catch and specific error ?

Yeah, that's my thinking.

**Comment by buchdag** [comment]
I think there is a general lack of clarity and consistence in the code of the `ChainFile`, `FullChainFile` and `FullFile` plugins methods.

`ChainFile`:
```
def load_from_content(self, content):
        chain = [self.load_cert(cert_data)
                 for cert_data in split_pems(content)]
        return self.Data(account_key=None, key=None, cert=None, chain=chain)

def save(self, data):
        return self.save_to_file(_PEMS_SEP.join(
            self.dump_cert(chain_cert) for chain_cert in data.chain))
```

`FullFile`:
```
def load_from_content(self, content):
        pems = split_pems(content)
        return self.Data(
            account_key=None,
            key=self.load_key(next(pems)),
            cert=self.load_cert(next(pems)),
            chain=[self.load_cert(cert) for cert in pems],
        )

def save(self, data):
    pems = [self.dump_key(data.key), self.dump_cert(data.cert)]
    pems.extend(self.dump_cert(cert) for cert in data.chain)
    self.save_to_file(_PEMS_SEP.join(pems))
```

Is there a reason not to code the two in a similar fashion for clarity sake ? For instance, `ChainFile` could be more like:

```
def load_from_content(self, content):
        pems = split_pems(content)
        return self.Data(
            account_key=None,
            key=None,
            cert=None,
            chain=[self.load_cert(cert) for cert in pems],
        )

def save(self, data):
        pems = [self.dump_cert(data.cert)]
        pems.extend(self.dump_cert(cert) for cert in data.chain)
        return self.save_to_file(_PEMS_SEP.join((pems))
```

`FullChainFile` is a subclass of `ChainFile`:
```
def load(self):
        data = super(FullChainFile, self).load()
        if data.chain is None:
            cert, chain = None, None
        else:
            cert, chain = data.chain[0], data.chain[1:]
        return self.Data(account_key=data.account_key, key=data.key,
                         cert=cert, chain=chain)

    def save(self, data):
        return super(FullChainFile, self).save(self.Data(
            account_key=data.account_key, key=data.key,
            cert=None, chain=([data.cert] + data.chain)))
```

I don't get what is gained by doing that, lt really look very contrived to me. Couldn't this just be similar to `FullFile` ? Am I completely missing the point ?

**Comment by zenhack** [comment]
Frankly, the whole IOPlugin hierarchy is needlessly complicated, and generally not very good code (I didn't write it...). You're not missing anything. Cleanup patches would certainly be welcome.

**Comment by buchdag** [comment]
Ok, good to know, I might start with that.

Do you think splitting some stuff out of `simple.py` into separate modules (at least the IOPlugin and the tests) might be possible too ?

**Comment by zenhack** [comment]
Yeah, I think that would make things easier to navigate.

**Comment by buchdag** [comment]
@zenhack you can close this if you want, that was fixed by #96

**Comment by zenhack** [comment]
Yep, thanks for pointing it out.
